pkg/configv2/migrate.go

```diff
@@ -0,0 +1,338 @@
+package configv2
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+
+	v1types "github.com/bacalhau-project/bacalhau/pkg/config/types"
+	"github.com/bacalhau-project/bacalhau/pkg/configv2/types"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+)
+
+func MigrateV1(in v1types.BacalhauConfig) (types.Bacalhau, error) {
+	publisherConfig, err := migratePublishers(in.Node)
+	if err != nil {
+		return types.Bacalhau{}, fmt.Errorf("migrating publisher config: %w", err)
+	}
+	executorConfig, err := migrateEngines(in.Node)
+	if err != nil {
+		return types.Bacalhau{}, fmt.Errorf("migrating executor config: %w", err)
+	}
+	out := types.Bacalhau{
+		DataDir: "~/.bacalhau",
+		// TODO(forrest) [review]: when migrating should the address come from the server or client when both are present?
+		API: types.API{
+			Address: "TODO",
+			TLS: types.TLS{
+				// TODO(forrest) [review]: when migrating if both the server and client have TLS configs, which do we take?
+				CertFile: "TODO",
+				KeyFile:  "TODO",
+				CAFile:   "TODO",
+			},
```

Comment:
@wdbaruni how should we address these TODOs?

Reply:
https://github.com/bacalhau-project/bacalhau/pull/4333#discussion_r1728025618